### PR TITLE
[Example] abstracts BucketizeSampler to be usable outside of HuBERT example.

### PR DESCRIPTION
**Pull request by erip**
This PR:

- Replaces the `data_source` with `lengths`
- Adds a `shuffle` argument to decide whether to shuffle the samples in the buckets.
- Add `max_len` and `min_len` to filter out samples that are > max_len or < min_len.

cc @nateanl

### CODE REVIEW

File: examples/hubert/dataset/hubert_dataset.py [hunk 1]
```diff
@@ -34,65 +33,77 @@ class BucketizeSampler(BatchSampler):
...
+
+        if not (0 <= min_len <= max_len):
+            raise AssertionError("``min_len`` should be non-negative and smaller than ``max_len``")
         if max_token_count is not None and batch_size is not None:
             raise AssertionError("The ``max_token_count`` and ``batch_size`` can't be both set.")
-        self.data_source = data_source
+        # Filter out examples which are outside the bounds of [min_len, max_len]
+        self.lengths = [length for length in lengths if min_len <= length <= max_len]
+        if len(self.lengths) == 0:
+            raise AssertionError("``lengths`` cannot be empty after filtering.")
```
Comment (nateanl) [inline examples/hubert/dataset/hubert_dataset.py:54]: ``lengths`` needs to be sorted and a new indices list is used for batching. Although bucketization already gathered the samples with similar lengths, sorting the lengths can help reduce the gap when transferring to a new bucket.

For example:
batch_size = 16
batches = [0:8] is from bucket 0
bucket[8:16] is from bucket 100
The length different will be large.

File: examples/hubert/dataset/hubert_dataset.py [hunk 2]
```diff
@@ -1,31 +1,30 @@
...
-from torch.utils.data import Dataset, BatchSampler
+from torch.utils.data import BatchSampler, Dataset
 
 
 class BucketizeSampler(BatchSampler):
     """Buketize sampler for data with different lengths to reduce number of paddings.
 
     Args:
-        data_source (Dataset): The dataset to sample
+        lengths (List[int]): The lengths of the examples in the dataset.e
```
Comment (nateanl) [inline examples/hubert/dataset/hubert_dataset.py:16]:
```suggestion
        lengths (List[int]): The lengths of the samples in the dataset.
```

File: examples/hubert/dataset/hubert_dataset.py [hunk 3]
```diff
@@ -34,65 +33,77 @@ class BucketizeSampler(BatchSampler):
...
 
-    def _get_buckets(self, data_source: Dataset, num_buckets: int) -> Dict[int, Tensor]:
+    def _get_buckets(self, lengths: List[int], num_buckets: int, min_len: int, max_len: int) -> Dict[int, Tensor]:
         """Generate buckets based on the dataset.
         Args:
-            data_source (Dataset): The dataset object to bucketize.
+            lengths (List[int]): The lengths of the examples in the dataset.e
             num_buckets (int): The number of buckets.
+            min_int (int): The lower bound of the evenly spaced length intervals to determine bucket width.
+            max_int (int): The upper bound of the evenly spaced length intervals to determine bucket width.
```
Comment (nateanl) [inline examples/hubert/dataset/hubert_dataset.py:66]:
```suggestion
            min_len (int): The lower bound of the evenly spaced length intervals to determine bucket width.
            max_len (int): The upper bound of the evenly spaced length intervals to determine bucket width.
```

**Review by nateanl — COMMENTED** [review]
LGTM! Could you address the comments and update the PR? Thanks.

File: examples/hubert/dataset/hubert_dataset.py (hunk 3 above)
Comment (erip) [reply examples/hubert/dataset/hubert_dataset.py:66]: Oops, good catch. :-)

File: examples/hubert/dataset/hubert_dataset.py [hunk 4]
```diff
@@ -34,65 +33,78 @@ class BucketizeSampler(BatchSampler):
...
+            max_len = max(lengths)
+
+        if not (0 <= min_len <= max_len):
+            raise AssertionError("``min_len`` should be non-negative and smaller than ``max_len``")
         if max_token_count is not None and batch_size is not None:
             raise AssertionError("The ``max_token_count`` and ``batch_size`` can't be both set.")
-        self.data_source = data_source
+        # Filter out examples which are outside the bounds of [min_len, max_len]
+        # sort to minimize gap when bucketizing.
+        self.lengths = list(sorted(length for length in lengths if min_len <= length <= max_len))
```
Comment (nateanl) [inline examples/hubert/dataset/hubert_dataset.py:53]: The indices needs to be updated since `lengths` order is changed.
```python
length_id = [(length, i) for i, length in enumerate(lengths) if min_len <= length <= max_len]
sorted_length_id = list(sorted(length_id, key = lambda x : x[0]))
self.lengths = ...
self.indices = ...
```

Comment (erip) [reply examples/hubert/dataset/hubert_dataset.py:53]: I see. So the `for i, _ in enumerate(len_list)` from earlier code should just be `for i in self.indices`?

Comment (erip) [reply examples/hubert/dataset/hubert_dataset.py:53]: Maybe to make it equivalent it would be `for i, _ in zip(self.indices, self.lengths)`, but the zip and ignore is basically a no-op.

Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:53]: Exactly. ``self.indices`` makes more sense. We just need to make sure the index matches the one from the original dataset.

Comment (erip) [reply examples/hubert/dataset/hubert_dataset.py:53]: I have something like this:

```patch
+        filtered_lengths = torch.tensor([length for length in lengths if min_len <= length <= max_len])
+        if len(filtered_lengths) == 0:
             raise AssertionError("``lengths`` cannot be empty after filtering.")
+        self.length_idx = torch.argsort(filtered_lengths)
+        self.lengths = filtered_lengths[length_idx]
         self.max_token_count = max_token_count
         self.batch_size = batch_size
-        self.buckets = self._get_buckets(self.lengths, num_buckets, min_len, max_len)
+        self.buckets = self._get_buckets(self.reordered_lengths, self.length_idx, num_buckets, min_len, max_len)
```

where `_get_buckets`s only code change is that the enumeration is replaced by iteration over `self.indices`. Basically it's keeping track of the original place of the example in the dataset. This seems pretty close to my implementation -- sorry it's taking a bit of time for this to click. 😅

Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:53]: Thanks. When we filter out the lengths, the indices already changed. So we can sort the lengths first to get the indices, then filter the lengths and indices, what do you think?

Comment (erip) [reply examples/hubert/dataset/hubert_dataset.py:53]: Yup, realized that right after posting this diff. OK, I'll make these changes quickly and push it. Looks like CI failures are unrelated, too...

**Review by nateanl — APPROVED** [review]
LGTM! Thanks for working on it.

**Comment by nateanl** [comment]
cc @ejguan This PR will be helpful when designing the bucketizeDataPipe.

**Comment by erip** [comment]
Lingering TODO is to fix the `HuBERTDataSet` to use this new interface. Do you want to take that @nateanl or shall I?

**Comment by nateanl** [comment]
Sure. I will do it. Testing the Dataset requires running the preprocessing, but since I have done that part it's easy to me.
